### ISSUE TITLE
move manifesto out of the store state

### ISSUE THREAD

**Issue by aeschylus** [ProjectMirador/mirador#1663]

**Comment by mejackreed** [comment]
@jbaiter brought up resisting looking at the implementation in https://github.com/stephenwf/iiif-redux

**Comment by jbaiter** [comment]
It does look like it would take care of quite a lot for us: https://github.com/stephenwf/iiif-redux/blob/master/CONCEPTS.md
The normalized state concept is pretty similar to what @charbugs and @christopher-johnson  have suggested, I think. And we have also looked at the `reselect` library already, which it uses, so it seems like it might be a good fit.

**Comment by mejackreed** [comment]
I've tried to implement it. One problem that I have run into so far is that the library is yet unreleased as a package. @stephenwf are there plans to release iiif-redux?

One other question I had was, to what extent could we extend it?

**Comment by stephenwf** [comment]
**tl;dr** it has extensions points for the guts, lets you add custom reducers, middleware and sagas so you can continue to use redux with it. Library need some API / usability work before release (soon).

I did some research a while back, to see what Mirador actually extracts from IIIF resources, to get an idea of what would be useful in terms of an API. This is my notes, some of the points on it I can't remember the context of them, but thought I'd chuck it in.

- Manifest
  - thumbnail
  - version
  - structures / ranges
  - _creating fake manifest wrappers_
  - _search within service (endpoint)_
  - label
  - viewing direction
- Collection
  - _version_
  - manifests
  - nested collections
- Annotations
  - Annotated by
  - _annotation endpoint (custom)_
  - _textual content_
  - tags
  - _selectors (box, svg, point)_
  - motivations

I've highlighted some gaps in italics that IIIF redux does not cover yet (and may not cover).

I also created a [frames extension](https://github.com/stephenwf/iiif-redux/blob/master/packages/iiif-redux/src/spaces/frames.js) internally to see how easy it was to extend with custom reducers/state and sagas. I think it went quite well. The frames demo was in part to test the waters on having multiple IIIF resources (and the state like current canvas, viewport etc) running at one time. 

The "frames" demo isn't aware of any user interface, but it is made to be extended itself. I created a proof of concept extensions around frames to mimic chrome tabs ([source](https://github.com/stephenwf/iiif-redux/blob/master/packages/iiif-redux-tabs/src/spaces/tabs.js) | [demo](https://iiif-redux-tabs.stephen.wf)) with some simple wrappers around the actions to create new actions more specific to the application (like createNewTab, updateSortOrder). 

That would be the sort of extension points for the guts of the library, if you chose to use it, extending and wrapping the primitive actions and selectors into specific concrete actions to match the viewer. 

As for extending with custom state/actions/sagas, the library does (or will?) export a `createStore` similar to Redux itself, so you can simple pass in state, reducers, sagas, middleware in the same way you use redux. The library "claims" a few paths in the state, but should be easy enough to work around. ([source](https://github.com/stephenwf/iiif-redux/blob/master/packages/iiif-redux/src/createStore.js#L9-L12))

I do want to release an initial version of the library, now that Presentation 3 Alpha has been implemented. I still need to bring some parity to the Presentation 2 and 3 APIs so you can use the same selectors on either resource, but I'm more than happy to expedite development of that if you'd like to prototype.

**Comment by christopher-johnson** [comment]
While I realize that there are concerns with implementing a server, I need to say that the `iiif-redux` approach seems very complex compared with GraphQL.  Apollo Server also supports paging and handles the remote async transport server side.  Basically, GraphQL allows you to provide criteria to get exactly what you need from a complex/deeply nested object, just like SQL.  For example, here is a query for a canvas using the `manifestId` and the `canvasId`.   
```graphql
query { canvas(manifestId: "https://iiif.bodleian.ox.ac.uk/iiif/manifest/d3a54f05-b2a7-4a3b-a181-fed761e41d67.json", 
canvasId: "https://iiif.bodleian.ox.ac.uk/iiif/canvas/d3a54f05-b2a7-4a3b-a181-fed761e41d67.json"),
            {id, label, width, height}
          }
```
which returns this object:
```json
{
  "data": {
    "canvas": {
      "id": "https://iiif.bodleian.ox.ac.uk/iiif/canvas/d3a54f05-b2a7-4a3b-a181-fed761e41d67.json",
      "label": "Whole page",
      "width": 5822,
      "height": 8544
    }
  }
}
```
The server requires a `resolver` method for the query type which is like reducer.  This is the resolver for a [canvas](https://github.com/ubl-chj/apollo-presentation-api/blob/master/packages/server/src/index.ts#L23-L27)

The task for the client becomes relatively simple.  Get specific values from JSON objects and pass them as properties to rendering components.  For me, it is not clear what all of the redux machinery actually brings to the table other than being able to serialize state.  I should also be possible to use redux and the apollo client together.

**Comment by stephenwf** [comment]
@christopher-johnson There is a proof of concept running GraphQL schema on the client side, as an abstraction around IIIF Redux and the selectors. Pull request is here: https://github.com/stephenwf/iiif-redux/pull/231

Some bits from the PR:
* [An example of making the graphql query on a component (fully functional)](https://github.com/stephenwf/iiif-redux/blob/f603921401caaa9f9e4bdfa201a60bce3e484577/packages/iiif-graphql/src/components/HelloWorld.js)
* [Generic GraphQL schema for IIIF (P3)](https://github.com/stephenwf/iiif-redux/blob/f603921401caaa9f9e4bdfa201a60bce3e484577/packages/iiif-graphql/src/schema.graphql)

There's definite potential in IIIF over GraphQL (server or client) and will really make developing with it much nicer. It is however either heavy on the client side, or requires a server. 

On the point of complexity, IIIF redux is very very verbose to use just now, but just like writing any GraphQL query, or SQL query, you write it once and use it many times. Abstract common sets of fields for resources and build up a library of selectors.

**Comment by christopher-johnson** [comment]
@stephenwf Nice work!  Interesting discussion on the PR as well.  Do we have sufficient motivation here to seek out a solution in the community to the server question?

**Comment by jbaiter** [comment]
> While I realize that there are concerns with implementing a server

I think there's more than just concerns, it's a complete deal breaker for most Mirador use cases.

**Comment by christopher-johnson** [comment]
> This completely contradicts the whole "Users should be able to run Mirador by dropping a <script> tag idea. 

These users have zero possibility to implement a full-featured read-write web annotation platform using this method.  Such a platform is one requirement of the Handschriftenportal.  So constraining the Mirador 3 project design to the most simple use case does not help us achieve that goal. 

I agree that it should be possible for Mirador 3 to work in a zero-configuration legacy mode.  But on the other hand, it should equally and seamlessly provide the extensible backend interfaces necessary to support complex distributed systems architecture.

> Also, how would keeping the bridge private (i.e. not accessible from the client's browser) work?

[Authentication](https://www.apollographql.com/docs/apollo-server/features/authentication.html).

**Comment by mejackreed** [comment]
There is some great discussion here with a lot of good information about various use cases and strategies.

While I agree that there are some strong architectural considerations on why a GraphQL backend can make our life easier for reading IIIF content, I'm concerned that takes us in a direction that we may not be prepared to handle, as a community of software maintainers and as a community of content providers. I think such a significant change (while already being adopted widely in tech) may be difficult to get initial traction within cultural heritage organizations which constitute the majority of Mirador users. Mirador has chosen the current IIIF as our interoperability platform of choice, and adding a new layer/gateway/bridge/external API to that is out of scope.

![screen shot 2019-01-24 at 7 31 58 am](https://user-images.githubusercontent.com/1656824/51684844-3f651000-1faa-11e9-9f75-81b7ccbf23ff.png)

A containing application can and should connect to additional API's. We shouldn't enforce Mirador users to do that though.

> implement a full-featured read-write web annotation platform 

Yes, Mirador3 can/will function as a viewer within an application that does this. But Mirador 3 cannot take on all of this, and it should handle a simpler scenario (which we are targeting in this first workcycle) while providing an interface for plugins and more complex scenarios.

> it should equally and seamlessly provide the extensible backend interfaces necessary to support complex distributed systems architecture

Sure! Though it might be more difficult at first, we need a client that can handle IIIF content first and foremost. I don't think we are closing off this possibility.

> I just don't think that this is within scope for the development of Mirador 3, and I thought we all agreed to that during the Kick-Off.

This is also my understanding of agreements from last week. Let's table the GraphQL discussion for now and move the discussion back to be more focused on the original intent of the issue created by @aeschylus ? 

For discussions around how Mirador could interact with additional API's beyond IIIF, please create a separate issue. Thanks everyone!

**Comment by stephenwf** [comment]
With [React hooks]() shipping on monday, I put together some quick examples of using IIIF Redux with it:

* [Some simple selections](https://github.com/stephenwf/iiif-redux/blob/580f22e403f67d4c1f0995cf238e895fcddacb9e/packages/iiif-redux-cookbook/src/components/HooksExample/HooksExample.js)
* [Minimum viable viewer](https://github.com/stephenwf/iiif-redux/blob/580f22e403f67d4c1f0995cf238e895fcddacb9e/packages/iiif-redux-cookbook/src/components/HookViewer/HookViewer.js)

Long and short of it is in functional components you can do something like:
```jsx
function RenderLabel() {
  const { label } = useManifest();

  return <div>{ t(label) }</div>;
}
```

Then when its a child of (however deep down) of the `Manifest` context, it will know to load that one.
```jsx
const App = (
  <Manifest id="...">
    <RenderLabel />
  </Manifest>
);
```

Nice loose coupling for creating UI that requires bits of IIIF state. It's all still exploratory, but would be keen to get some feedback.
